net/WebSocketHandler.hpp

```diff
@@ -599,48 +639,89 @@ class WebSocketHandler : public ProtocolHandlerInterface
         if (!socket->isWebSocket())
         {
             LOG_WRN("Attempted ping on non-upgraded websocket! #" << socket->getFD());
-            _lastPingSentTime = now; // Pretend we sent it to avoid timing out immediately.
+            setPingPongTime(now); // Pretend we sent it to avoid timing out immediately.
             return;
         }
-
-        LOGA_TRC(WebSocket, "Sending " << (const char*)(code == WSOpCode::Ping ? " ping" : "pong"));
+        if (code == WSOpCode::Ping)
+        {
+            ++_openPingCount;
+            LOGA_TRC(WebSocket, "Sending ping, open pings left " << _openPingCount);
+        }
+        else
+        {
+            LOGA_TRC(WebSocket, "Sending pong");
+        }
+        _lastPingSentTime = now;
         // FIXME: allow an empty payload.
         sendMessage(data, len, code, false);
-        _lastPingSentTime = now;
     }
 
 public:
+    /// Sends a native control-frame ping message
     void sendPing(std::chrono::steady_clock::time_point now,
                   const std::shared_ptr<StreamSocket>& socket)
     {
-//        assert(!_isClient);
+        if (_isClient)
+            LOG_DBG("Clients should not send pings, only servers");
+//        assert(_isClient);
         sendPingOrPong(now, "", 1, WSOpCode::Ping, socket);
     }
+    #if 0
+        /// Sends a native control-frame ping message
+        /// Only for debugging purposes. NOT WORKING (SOCKET THREAD AFFINITY)
+        void sendPing(std::chrono::steady_clock::time_point now)
+        {
+            std::shared_ptr<StreamSocket> socket = getSocket().lock();
+            if (socket) {
+                sendPing(now, socket);
+            } else {
+                LOG_DBG("sendPing: Socket n/a yet");
+            }
+        }
+    #endif
 
+    /// Sends a native control-frame pong message
     void sendPong(std::chrono::steady_clock::time_point now,
                   const char* data, const size_t len,
                   const std::shared_ptr<StreamSocket>& socket)
     {
+        if (!_isClient)
+            LOG_WRN("Servers should not send pongs, only clients");
+//        assert(!_isClient);
         sendPingOrPong(now, data, len, WSOpCode::Pong, socket);
     }
 #endif
 
     /// Do we need to handle a timeout ?
-    void checkTimeout([[maybe_unused]] std::chrono::steady_clock::time_point now) override
+    bool checkTimeout([[maybe_unused]] std::chrono::steady_clock::time_point now) override
     {
 #if !MOBILEAPP
         if (_isClient)
-            return;
+        {
+            return false;
+        }
 
-        const auto timeSincePingMicroS
-            = std::chrono::duration_cast<std::chrono::microseconds>(now - _lastPingSentTime);
-        if (timeSincePingMicroS >= PingFrequencyMicroS)
+        if (_openPingCount > 0 && _pingTimeout.count() > std::numeric_limits<double>::epsilon() &&
+            _pingMicroS.average() >= _pingTimeout.count())
```

Comment:
I'm not sure I follow why the ping timeout needs to figure out an average.

Timeouts are absolute. Ping timeout is no different.

If the idea is to have some tolerance, the traditional/typical way of doing that is simple: have a soft limit and a hard limit. Easiest and most common implementation considers the first timeout to be the soft limit, issues a warning, and sends another ping. If that second ping also times out (i.e. it has been 2x timeout-duration since last message from the peer), then the hard limit is reached and the connection is interrupted.

If there is a deeper explanation why we need such complexity, a detailed comment is warranted. But a simple soft/hard limit timeout is sufficient (in fact, one can argue that averages, weighted or not, are very problematic and highly unpredictable).

Reply:
Time weighted values (here for the ping-duration) is most stable here, since we don't like to pull the plug on a 1-2 off hit. It allows to use the integrated value over time.
Also checked for a more simple exponential or mixed (last one with more weight) .. while that is better than just the last value, it still could give a disconnect criteria for 'just a bad few seconds'.
All this assuming that a false disconnect is to be avoided.

I couldn't produce good test values with firefox, but saw a range of 20-120us even locally (6x). 
Ranges we expects are ~1ms - 2s AFAIK.